examples/navier-stokes/common.h

```diff
@@ -113,6 +169,39 @@ CEED_QFUNCTION(Setup)(void *ctx, CeedInt Q,
   return 0;
 }
 
+CEED_QFUNCTION(Setup2d)(void *ctx, CeedInt Q,
+                      const CeedScalar *const *in, CeedScalar *const *out) {
+  // Inputs
+  const CeedScalar (*J)[2][Q] = (CeedScalar(*)[2][Q])in[0],
```

Comment:
Do we care about making this declaration and the other similar ones as 
```
const CeedScalar (*J)[2][Q] = (const CeedScalar(*)[2][Q])in[0]
```
to avoid the warnings on intel compilers?

Reply:
Better to be strict to avoid the warnings (even if they aren't useful). Are there many such instances? (GCC used to warn, but decided to switch to a more useful interpretation of the standard.)

Reply:
Oh yes, there are many. We have a lot of these declarations using VLA in our header files:
```
In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(46):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/common.h(120): warning #144: a value of type "CeedScalar={double} (*)[3][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[3][Q]"
    const CeedScalar (*J)[3][Q] = (CeedScalar(*)[3][Q])in[0],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(46):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/common.h(175): warning #144: a value of type "CeedScalar={double} (*)[2][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[2][Q]"
    const CeedScalar (*J)[2][Q] = (CeedScalar(*)[2][Q])in[0],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(46):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/common.h(220): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*u)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(107): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*X)[Q] = (CeedScalar(*)[Q])in[0];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(150): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(151): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(152): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[2],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(153): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*x)[Q] = (CeedScalar(*)[Q])in[3];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(267): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(268): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(269): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdot)[Q] = (CeedScalar(*)[Q])in[2],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(270): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[3];
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(89): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*X)[Q] = (CeedScalar(*)[Q])in[0];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(132): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(133): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(134): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[2],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(135): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*x)[Q] = (CeedScalar(*)[Q])in[3];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(224): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(225): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(226): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdot)[Q] = (CeedScalar(*)[Q])in[2],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(227): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[3];
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(132): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*X)[Q] = (CeedScalar(*)[Q])in[0];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(209): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(210): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(211): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[2],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(212): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*x)[Q] = (CeedScalar(*)[Q])in[3];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(461): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(462): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(463): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdot)[Q] = (CeedScalar(*)[Q])in[2],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(464): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[3];
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(50):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent_primitive.h(134): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*X)[Q] = (CeedScalar(*)[Q])in[0];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(50):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent_primitive.h(213): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(50):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent_primitive.h(214): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(50):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent_primitive.h(215): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdot)[Q] = (CeedScalar(*)[Q])in[2],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(50):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent_primitive.h(216): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[3];
                                   ^


```